### PR TITLE
Feat/track details itunes

### PR DESCRIPTION
**Pull request by abhishek-ws**
### Ticket Link

---

### Related Links

---

### Description

This PR contains following updates in this application :-
- Every song card can be clicked and user will be sent to the individual route of that song called TrackDetails page.
- The Redux store maintains the cache storage of every details of the songs user clicked and it is responsible to check and dispatch API call if the trackDetails are not available inside cache.
- More UI Updates in both Home and TrackDetails component.

### Steps to Reproduce / Test
yarn test

### GIF's
-  #### UI Look HomePage and TrackDetails page
PR3_UIDemo

-  #### Redux State management
PR3_ReduxDemo

-  #### Output of yarn test
yarnTest3

- ### Responsive Home Page Grid Demo
PR3-ResponsivenessUIdemo

- ### Responsive Track Details Page
PR3-ResponsivenessUITrackDetailsDemo

### CODE REVIEW

File: app/components/SongCard/index.js [hunk 1]
```diff
@@ -10,7 +10,8 @@ import styled from 'styled-components';
 import { Card, Typography, Button } from 'antd';
 import { PlayCircleTwoTone, StopTwoTone } from '@ant-design/icons';
 import { colors } from '@app/themes';
-import { translate } from '@components/IntlGlobalProvider';
+import { translate } from '../IntlGlobalProvider/index';
```
Comment (alichherawalla) [inline app/components/SongCard/index.js:13]:
```
import { translate } from '@components/IntlGlobalProvider';
```

Comment (alichherawalla) [reply app/components/SongCard/index.js:13]: // third party
// alias
// relative

File: app/components/SongCard/index.js [hunk 2]
```diff
@@ -124,7 +126,9 @@ export function SongCard({ song, trackDetails, width, height, onActionClick }) {
     <Container width={width} height={height} data-testid="song-card">
       <HeaderFooter>
         <StyledImage src={artworkUrl100} />
-        <StyledT trackDetails={trackDetails} text={trackName} />
+        <Link to={`/details/${trackId}`}>
+          <StyledT trackDetails={trackDetails} text={trackName} />
+        </Link>
```
Comment (alichherawalla) [inline app/components/SongCard/index.js:131]: songElement?.current?.paused

Comment (alichherawalla) [reply app/components/SongCard/index.js:131]:
```
const getPlayState = () => songElement?.current ? songElement.current.paused : false 
```

File: app/components/SongCard/index.js [hunk 3]
```diff
@@ -171,8 +175,6 @@ SongCard.propTypes = {
   width: PropTypes.number,
```
Comment (alichherawalla) [inline app/components/SongCard/index.js:175]: add song shape in the propTypes

File: app/components/SongCard/tests/index.test.js [hunk 4]
```diff
@@ -19,24 +36,84 @@ describe('<SongCard/> tests', () => {
   };
 
   it('should render and match the snapshot', () => {
-    const { baseElement } = renderWithIntl(<SongCard song={song} />);
+    const { baseElement } = renderWithIntl(<SongWrapper song={song} />);
```
Comment (alichherawalla) [inline app/components/SongCard/tests/index.test.js:39]: inside renderWithIntl replace ```children ``` with
```
<BrowserRouter>{children}</BrowserRouter>
```

File: app/containers/ItunesContainer/TrackDetails/index.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        setCurrentTrack(audioRef);
+      }
+    }
+  };
+
+  return (
+    <Container>
+      <If condition={!error} otherwise={<T data-testid="track-detail-error" id="something_went_wrong" />}>
+        <Skeleton data-testid="skeleton-card" loading={isEmpty(trackDetails)} active>
+          <SongCard
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/TrackDetails/index.js:81]:
```
 <SongCard
            song={trackDetails}
            trackDetails
            width={width}
            height={height}
            padding={padding}
            onActionClick={handleActionClick}
          />
``

File: app/containers/ItunesContainer/TrackDetails/tests/index.test.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
+/**
+ *
+ * Tests for TrackDetails
+ *
+ *
+ */
+
+import React from 'react';
+import { timeout, renderProvider } from '@utils/testUtils';
+// import { fireEvent } from '@testing-library/dom'
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/TrackDetails/tests/index.test.js:10]: remove commented code

File: app/containers/ItunesContainer/TrackDetails/tests/index.test.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    await timeout(500);
+    expect(submitSpyClearTrackDetails).toBeCalled();
+    expect(submitSpyTrackSearch).toBeCalled();
+  });
+
+  it('should validate mapDispatchToProps actions', async () => {
+    const dispatchSearchSpy = jest.fn();
+    const trackId = 1234;
+    const actions = {
+      dispatchTrackSearch: { trackId: 1234, type: 'SEARCH_TRACK' },
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/TrackDetails/tests/index.test.js:62]: dispatchTrackSearch: { trackId, type: <Type.SEARCH_TRACK> },

File: app/containers/ItunesContainer/saga.js [hunk 8]
```diff
@@ -1,9 +1,13 @@
...
+import { getSongs, getTrackDetails } from '@services/songsApi';
+import { selectSongsCache } from './selectors';
+import { isEmpty } from 'lodash';
 
-const { SEARCH_ITUNES } = itunesContainerTypes;
-const { successSearchItunes, failureSearchItunes } = itunesContainerCreators;
+const { SEARCH_ITUNES, SEARCH_TRACK } = itunesContainerTypes;
+const { successSearchItunes, failureSearchItunes, successSearchTrack, failureSearchTrack } = itunesContainerCreators;
+
+// const songsCache = selectSongsCache();
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/saga.js:10]: remove

File: app/containers/ItunesContainer/reducer.js [hunk 9]
```diff
@@ -41,6 +50,24 @@ export const itunesContainerReducer = (state = initialState, action) =>
...
+      case itunesContainerTypes.SEARCH_TRACK:
+        draft.trackSearchError = null;
+        draft.trackDetails = {};
+        draft.trackId = action.trackId;
+        break;
+      case itunesContainerTypes.SUCCESS_SEARCH_TRACK:
+        draft.trackDetails = {};
+        draft.trackSearchError = null;
+        draft.songsCache[draft.trackId] = action.data;
+        draft.trackDetails = draft.songsCache[draft.trackId].results[0];
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/reducer.js:62]: move compute to saga

File: app/containers/ItunesContainer/reducer.js [hunk 10]
```diff
@@ -41,6 +50,24 @@ export const itunesContainerReducer = (state = initialState, action) =>
...
+        break;
+      case itunesContainerTypes.FAILURE_SEARCH_TRACK:
+        draft.trackDetails = {};
+        draft.trackSearchError = get(action.error, 'message', translate('something_went_wrong'));
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/reducer.js:66]: move this to the saga

File: app/containers/ItunesContainer/tests/index.test.js [hunk 11]
```diff
@@ -53,7 +63,47 @@ describe('ItunesContainer Tests', () => {
...
+  it('should render the error message', async () => {
+    const { getByTestId } = renderProvider(
+      <ItunesWrapper dispatchSearchSongs={mockDispatchSearch} searchError="error" />
+    );
+
+    await timeout(500);
+    expect(getByTestId('itunes-error-message')).toBeInTheDocument();
+  });
+
+  it('should render the default message', async () => {
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/tests/index.test.js:103]: write a test to make sure that if 50 tracks are passsed, 50 trackCards are rendered

Comment (alichherawalla) [inline app/containers/ItunesContainer/tests/index.test.js:103]: write a test to make sure if one card is played, the rest are paused

File: app/utils/returnRouter.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { StaticRouter } from 'react-router';
+import PropTypes from 'prop-types';
```
Comment (alichherawalla) [inline app/utils/returnRouter.js:3]: remove this

File: app/containers/ItunesContainer/tests/index.test.js (hunk 11 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/tests/index.test.js:103]: Done

File: app/components/SongCard/tests/index.test.js (hunk 4 above)
Comment (abhishek-ws) [reply app/components/SongCard/tests/index.test.js:39]: Applied

File: app/components/SongCard/index.js (hunk 3 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:175]: Added

File: app/containers/ItunesContainer/TrackDetails/tests/index.test.js (hunk 7 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/tests/index.test.js:62]: Updated

File: app/containers/ItunesContainer/TrackDetails/tests/index.test.js (hunk 6 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/tests/index.test.js:10]: Removed accordingly

File: app/containers/ItunesContainer/TrackDetails/index.js (hunk 5 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/index.js:81]: Updated

File: app/components/SongCard/index.js (hunk 1 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:13]: Updated

File: app/containers/ItunesContainer/saga.js (hunk 8 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/saga.js:10]: Done

File: app/containers/ItunesContainer/reducer.js (hunk 9 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/reducer.js:62]: Done

File: app/utils/returnRouter.js (hunk 12 above)
Comment (abhishek-ws) [reply app/utils/returnRouter.js:3]: Done

File: app/components/SongCard/index.js [hunk 13]
```diff
@@ -90,10 +87,22 @@ const StyledImage = styled.img`
...
 `;
 
+const StyledPlayIcon = styled(PlayCircleTwoTone)`
+  && {
+    fontsize: 18;
+  }
+`;
+const StyledStopIcon = styled(StopTwoTone)`
+  && {
+    fontsize: 18;
```
Comment (alichherawalla) [inline app/components/SongCard/index.js:97]: this should be font-size right?

File: app/components/SongCard/index.js [hunk 14]
```diff
@@ -90,10 +87,22 @@ const StyledImage = styled.img`
...
+  }
+`;
+
 export function SongCard({ song, trackDetails, width, height, onActionClick }) {
-  const { trackName, trackPrice, artworkUrl100, previewUrl } = song;
-  const [play, setPlay] = useState(false);
+  const { trackName, trackPrice, artworkUrl100, previewUrl, trackId } = song;
+
   const songElement = useRef(null);
+  const [play, setPlay] = useState(false);
```
Comment (alichherawalla) [inline app/components/SongCard/index.js:105]: how come we're using state again ?

File: app/components/SongCard/index.js (hunk 3 above)
Comment (alichherawalla) [reply app/components/SongCard/index.js:190]: nice

File: app/containers/ItunesContainer/TrackDetails/index.js [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  height,
+  padding
+}) {
+  const { trackId } = useParams();
+
+  const [error, setError] = useState(false);
+
+  useEffect(() => {
+    if (!isEmpty(trackSearchError)) {
+      setError(true);
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/TrackDetails/index.js:55]: we don't need setError and error right?

File: app/containers/ItunesContainer/TrackDetails/tests/index.test.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    const { getByTestId } = renderWithIntl(
+      <TrackDetails
+        dispatchTrackSearch={submitSpyTrackSearch}
+        dispatchClearTrackDetails={submitSpyClearTrackDetails}
+        trackSearchError="error"
+      />
+    );
+    await timeout(500);
+    expect(getByTestId('track-detail-error')).toBeInTheDocument();
+  });
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/TrackDetails/tests/index.test.js:69]: add tests to make sure that number of SongCards rendered is equal to number of tracks sent as props

File: app/containers/ItunesContainer/saga.js [hunk 17]
```diff
@@ -1,20 +1,49 @@
...
+    const response = { data, trackDetails: data.results[0] };
+    yield put(successSearchTrack(response));
+  } else {
+    const res = yield call(getTrackDetails, action.trackId);
+    const { data, ok } = res;
+    if (ok) {
+      const response = { data, trackDetails: data.results[0] };
+      yield put(successSearchTrack(response));
+    } else {
+      const error = data?.originalError?.message ?? translate('something_went_wrong');
```
Comment (alichherawalla) [inline app/containers/ItunesContainer/saga.js:40]: nice

File: app/components/SongCard/index.js (hunk 13 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:97]: Yes, fixed accordingly

File: app/containers/ItunesContainer/TrackDetails/index.js (hunk 15 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/index.js:55]: Updated in the code

File: app/components/SongCard/index.js (hunk 14 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:105]: Resolved

File: app/components/SongCard/index.js [hunk 18]
```diff
@@ -55,14 +52,14 @@ const ControlButton = styled(Button)`
 
 const StyledT = styled(T)`
   && {
-    font-size: ${(props) => (props.trackDetails ? 2 : 1.1)}em;
-    margin: ${(props) => (props.trackDetails ? '20 0' : 0)};
+    font-size: ${(props) => (props.full ? 2 : 1.1)}em;
+    margin: ${(props) => (props.full ? '20 0' : 0)};
```
Comment (vishnu-wednesday) [inline app/components/SongCard/index.js:56]: 20 what?

File: app/components/SongCard/index.js [hunk 19]
```diff
@@ -4,22 +4,23 @@
...
 
-import React, { useState, useRef, useEffect } from 'react';
+import React, { useRef, useState } from 'react';
 import PropTypes from 'prop-types';
 import styled from 'styled-components';
 import { Card, Typography, Button } from 'antd';
 import { PlayCircleTwoTone, StopTwoTone } from '@ant-design/icons';
 import { colors } from '@app/themes';
-import { translate } from '@components/IntlGlobalProvider';
+import { translate } from '@components/IntlGlobalProvider/';
```
Comment (vishnu-wednesday) [inline app/components/SongCard/index.js:13]: why the trailing '/'

File: app/components/SongCard/index.js [hunk 20]
```diff
@@ -90,44 +87,46 @@ const StyledImage = styled.img`
   margin-bottom: 2em;
 `;
 
+const StyledPlayIcon = styled(PlayCircleTwoTone)`
+  && {
+    font-size: 18;
```
Comment (vishnu-wednesday) [inline app/components/SongCard/index.js:92]: Missing unit?

File: app/components/SongCard/index.js [hunk 21]
```diff
@@ -90,44 +87,46 @@ const StyledImage = styled.img`
...
+  }
+`;
+const StyledStopIcon = styled(StopTwoTone)`
+  && {
+    font-size: 18;
```
Comment (vishnu-wednesday) [inline app/components/SongCard/index.js:97]: Here also

File: app/components/SongCard/index.js [hunk 22]
```diff
@@ -138,41 +137,52 @@ export function SongCard({ song, trackDetails, width, height, onActionClick }) {
       <IconsContainer>
         <ControlButton
           data-testid="play-btn"
-          onClick={() => handleMusic(actions.PLAY, previewUrl)}
-          disabled={play}
-          type={play ? 'text' : 'ghost'}
-          icon={<PlayCircleTwoTone style={commonIconStyle} />}
+          onClick={handleMusic}
+          disabled={isPlaying()}
+          type={isPlaying() ? 'text' : 'ghost'}
```
Comment (vishnu-wednesday) [inline app/components/SongCard/index.js:142]: What is this type?

File: app/components/SongCard/index.js (hunk 2 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:126]: Added

File: app/components/SongCard/tests/index.test.js [hunk 23]
```diff
@@ -28,15 +36,81 @@ describe('<SongCard/> tests', () => {
     expect(getAllByTestId('song-card').length).toBe(1);
   });
 
-  it('should play the song when "Play Button" clicked', () => {
+  it('should render No-price translation when trackPrice is not available', () => {
+    const song = {
```
Comment (vishnu-wednesday) [inline app/components/SongCard/tests/index.test.js:40]: maybe you can use the song object defined earlier. change it to let and modify it by using `song = {...song, anynewproperty}`

File: app/containers/ItunesContainer/TrackDetails/index.js [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ */
+
+import React, { useEffect, memo } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { injectIntl } from 'react-intl';
+import { useParams } from 'react-router';
+import { Skeleton } from 'antd';
+import SongCard from '@app/components/SongCard/index';
```
Comment (vishnu-wednesday) [inline app/containers/ItunesContainer/TrackDetails/index.js:13]: `@component/SongCard`

File: app/containers/ItunesContainer/TrackDetails/index.js [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+import styled from 'styled-components';
+import T from '@app/components/T/index';
```
Comment (vishnu-wednesday) [inline app/containers/ItunesContainer/TrackDetails/index.js:15]: Same as above

File: app/containers/ItunesContainer/TrackDetails/index.js [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+import { createStructuredSelector } from 'reselect';
+import { compose } from 'redux';
+import { injectSaga } from 'redux-injectors';
+import { selectItunesContainer, selectTrackDetails, selectTrackSearchError } from '../selectors';
+import itunesContainerSaga from '../saga';
+import { itunesContainerCreators } from '../reducer';
+import { Link } from 'react-router-dom';
+import { colors } from '@app/themes';
+import If from '@components/If/index';
```
Comment (vishnu-wednesday) [inline app/containers/ItunesContainer/TrackDetails/index.js:24]: `@component/If`

File: app/containers/ItunesContainer/TrackDetails/index.js [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+import { isEmpty } from 'lodash';
```
Comment (vishnu-wednesday) [inline app/containers/ItunesContainer/TrackDetails/index.js:25]: Also nitpick. Structure your imports better. 
// third party
// alias
// relative

Will make it easier to read it I guess

File: app/containers/ItunesContainer/index.js [hunk 28]
```diff
@@ -96,6 +96,7 @@ export function ItunesContainer({
 
   const handleOnChange = (searchTerm) => {
     if (!isEmpty(searchTerm)) {
+      // console.log('Something');
```
Comment (vishnu-wednesday) [inline app/containers/ItunesContainer/index.js:99]: remove

File: app/components/SongCard/index.js (hunk 20 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:92]: by default it takes px right, so was considering this

File: app/components/SongCard/index.js (hunk 21 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:97]: by default it takes px right, so was considering this

File: app/containers/ItunesContainer/TrackDetails/index.js (hunk 27 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/index.js:25]: Updated

File: app/containers/ItunesContainer/index.js (hunk 28 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/index.js:99]: Done

File: app/containers/ItunesContainer/TrackDetails/index.js (hunk 26 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/index.js:24]: Changed accordingly

File: app/containers/ItunesContainer/TrackDetails/index.js (hunk 25 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/index.js:15]: Changed accordingly

File: app/containers/ItunesContainer/TrackDetails/index.js (hunk 24 above)
Comment (abhishek-ws) [reply app/containers/ItunesContainer/TrackDetails/index.js:13]: Done

File: app/components/SongCard/index.js (hunk 22 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:142]: Button type, that changes according to disable, enable states when a song is played or stopped.

File: app/components/SongCard/index.js (hunk 21 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:97]: Changed accordingly

File: app/components/SongCard/index.js (hunk 20 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:92]: This is smomething,

File: app/components/SongCard/tests/index.test.js (hunk 23 above)
Comment (abhishek-ws) [reply app/components/SongCard/tests/index.test.js:40]: Updated accordigly

File: app/components/SongCard/index.js (hunk 20 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:92]: fixed the unit issue

File: app/components/SongCard/index.js (hunk 19 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js]: Changed accordingly

File: app/components/SongCard/index.js (hunk 18 above)
Comment (abhishek-ws) [reply app/components/SongCard/index.js:56]: Updated

**Comment by vishnu-wednesday** [comment]
screen_abhi

I have a small console error. Happens the first time I play something. The function `handleActionClick` is not tested too I believe. 

Also, could you maybe work a little on the responsivness?

**Comment by abhishek-ws** [comment]
> screen_abhi
> 
> I have a small console error. Happens the first time I play something. The function `handleActionClick` is not tested too I believe.
> Also, could you maybe work a little on the responsivness?

Fixed the issue  of console error on clicking the Play button accordingly,

**Comment by vishnu-wednesday** [comment]
Hmm... That is usually not how large projects will have media queries. That amount of media queries and breakpoints will inevitably lead to unmaintainable styles. 

If you look at `themes/media.js` you will see a small media query implementation. Use that. better yet use flexbox or grid to achieve a nice layout without media queries.

**Comment by vishnu-wednesday** [comment]
Looks good. I am not a big fan of all the media queries or components having rigid height and width. I feel we can move on though. 
Please feel free to merge. Also if you are interested 
https://css-tricks.com/beyond-media-queries-using-newer-html-css-features-for-responsive-designs/

**Pull request merged**